File: sort/quick-sort.go
package main

import "fmt"

func swap(a *int, b *int) {
	t := *a
	*a = *b
	*b = t
}

func partition(array []int, left int, right int) int {
	pivot := array[right]

	i := left - 1

	for j := left; j < right; j++ {
		if array[j] <= pivot {
			i++

			swap(&array[i], &array[j])
		}
	}
	swap(&array[i+1], &array[right])
	return i + 1
}

func quickSort(array []int, left int, right int) {
	if left < right {
		pi := partition(array, left, right)

		quickSort(array, left, pi-1)
		quickSort(array, pi+1, right)
	}
}

func main() {
	data := []int{8, 7, 2, 1, 0, 9, 6}
	dataLength := len(data)

	fmt.Println("unsorted array: ", data)

	quickSort(data, 0, dataLength-1)
	fmt.Println("sorted array: ", data)
}

File: sort/quick-sort.rs
fn partition(arr: &mut [i32]) -> usize {
    let pivot_index = arr.len() / 2;
    arr.swap(pivot_index, arr.len() - 1);

    let mut i = 0;
    for j in 0..arr.len() - 1 {
        if arr[j] <= arr[arr.len() - 1] {
            arr.swap(i, j);
            i += 1
        }
    }

    arr.swap(i, arr.len() - 1);
    return i;
}

fn quick_sort(arr: &mut [i32]) {
    if arr.len() < 1 {
        return;
    }
    let pivot_index = partition(arr);

    quick_sort(&mut arr[0..pivot_index]);
    quick_sort(&mut arr[pivot_index + 1..])
}

fn main() {
    let mut unsorted_array = vec![12, 11, 13, 5, 6, 2, -1, 4, -21, 34, 21, 38, 22];
    quick_sort(&mut unsorted_array);
    println!("{:?}", unsorted_array);
}
